gflownet/utils/batch.py

```diff
@@ -0,0 +1,346 @@
+                self.steps.append(env.n_actions)
+                self.masks_invalid_actions_forward.append(
+                    env.get_mask_invalid_actions_forward()
+                )
+                if self.loss == "flowmatch":
+                    parents, parents_a = env.get_parents(action=action)
+                    assert (
+                        action in parents_a
+                    ), f"""
+                    Sampled action is not in the list of valid actions from parents.
+                    \nState:\n{env.state}\nAction:\n{action}
+                    """
+                    self.all_possible_parents.append(parents)
+                    self.all_possible_parents_actions.append(parents_a)
+                if self.loss == "trajectorybalance":
+                    self.masks_invalid_actions_backward.append(
+                        env.get_mask_invalid_actions_backward(
+                            env.state, env.done, [action]
+                        )
+                    )
+            else:
+                if env.done:
+                    self.states.append(env.state)
+                    self.env_ids.append(env.id)
+                    self.steps.append(env.n_actions)
+
+    def process_batch(self):
+        """
+        Process internal lists to more convenient formats:
+        - converts and stacks list to a single torch tensor
+        - computes trajectory indicies (indecies of the states in self.states corresponding to each trajectory)
+        - if needed, computes states and parents in policy formats (stored in self.states_policy, self.parents_policy)
+        """
+        self.env_ids = tlong(self.env_ids, device=self.device)
+        self._process_states()
+        self.steps = tlong(self.steps, device=self.device)
+        self._process_trajectory_indices()
+        # process other variables, if we are in the train mode and recorded them
+        if len(self.actions) > 0:
+            self.actions = tfloat(
+                self.actions, device=self.device, float_type=self.float
+            )
+            self.done = tbool(self.done, device=self.device)
+            self.masks_invalid_actions_forward = tbool(
+                self.masks_invalid_actions_forward, device=self.device
+            )
+            if self.loss == "flowmatch":
+                self.all_possible_parents_state_idx = tlong(
+                    sum(
+                        [
+                            [idx] * len(p)
+                            for idx, p in enumerate(self.all_possible_parents)
+                        ],
+                        [],
+                    ),
+                    device=self.device,
+                )
+                self.all_possible_parents_actions = torch.cat(
+                    [
+                        tfloat(x, device=self.device, float_type=self.float)
+                        for x in self.all_possible_parents_actions
+                    ]
+                )
```

Comment:
I think that, for efficiency reasons, we should concatenate first and then call `tfloat`, to avoid creating many unnecessary tensors (inside `tfloat`). Tagging @carriepl to confirm that this would be better.

Reply:
I have re-implemented it [here](https://github.com/alexhernandezgarcia/gflownet/blob/batch_class-ahg/gflownet/utils/batch.py#L166)

Reply:
I wouldn't expect a big difference when running on CPU but it would definitely help when running on GPU (because of the latency associated with transfering each individual value VS the whole thing all at once).

Reply:
Done (here)[https://github.com/alexhernandezgarcia/gflownet/pull/111/commits/d3631459ec9ae11edfb1253cc58fd0d6eb71aef5]